stingray/deadtime/filters.py

```diff
@@ -0,0 +1,161 @@
+            mask[i] = False
+    return mask
+
+
+def _non_paralyzable_dead_time(event_list, dead_time):
+    event_list_dbl = (event_list - event_list[0]).astype(np.double)
+    dead_time_end = event_list_dbl + np.double(dead_time)
+    mask = np.ones(event_list_dbl.size, dtype=bool)
+    mask = _nonpar_core(event_list_dbl, dead_time_end, mask)
+    return event_list[mask], mask
+
+
+class DeadtimeFilterOutput(object):
+    uf_events = None
+    is_event = None
+    deadtime = None
+    mask = None
+    bkg = None
+
+
+def filter_for_deadtime(event_list, deadtime, bkg_ev_list=None,
+                        dt_sigma=None, paralyzable=False,
+                        additional_data=None, return_all=False):
+    """Filter an event list for a given dead time.
+
+    Parameters
+    ----------
+    ev_list : array-like or class:`stingray.events.EventList`
+        The event list
+    deadtime: float
+        The (mean, if not constant) value of deadtime
+
+    Other Parameters
+    ----------------
+    bkg_ev_list : array-like
+        A background event list that affects dead time
+    dt_sigma : float
+        The standard deviation of a non-constant dead time around deadtime.
+    return_all : bool
+        If True, return the mask that filters the input event list to obtain
+        the output event list.
+
+    Returns
+    -------
+    new_ev_list : class:`stingray.events.EventList` object
+        The filtered event list
+    additional_output : dict
+        Object with all the following attributes. Only returned if
+        `return_all` is True
+    uf_events : array-like
+        Unfiltered event list (events + background)
+    is_event : array-like
+        Boolean values; True if event, False if background
+    deadtime : array-like
+        Dead time values
+    bkg : array-like
+        The filtered background event list
+    mask : array-like, optional
+        The mask that filters the input event list and produces the output
+        event list.
+
+    """
+    additional_output = DeadtimeFilterOutput()
+    if isinstance(event_list, EventList):
+        ev_list = event_list.time
+        mjdref = event_list.mjdref
+    else:
+        ev_list = event_list
+        mjdref = 0
+
+    if deadtime <= 0.:
+        if deadtime < 0:
+            warnings.warn("Dead time < 0")
+        return event_list
+
+    # Create the total lightcurve, and a "kind" array that keeps track
+    # of the events classified as "signal" (True) and "background" (False)
+    if bkg_ev_list is not None:
+        tot_ev_list = np.append(ev_list, bkg_ev_list)
+        ev_kind = np.append(np.ones(len(ev_list), dtype=bool),
+                            np.zeros(len(bkg_ev_list), dtype=bool))
+        order = np.argsort(tot_ev_list)
+        tot_ev_list = tot_ev_list[order]
+        ev_kind = ev_kind[order]
+        del order
+    else:
+        tot_ev_list = ev_list
+        ev_kind = np.ones(len(ev_list), dtype=bool)
```

Comment:
I think for future potential work in terms of changes/debugging, it would be really helpful to have some comments in this function. It gets a bit confusing with the different event lists and `ev_kind` and `all_ev_kind` that having a few comments in there to talk through the logic would be super helpful.